### PR TITLE
Fix byte encoding, bypassing the vocabulary

### PR DESCRIPTION
**Pull request by matt-gardner**
Byte encoding converts tokens to ints directly, and using the vocabulary to then map those ints to other ints is both unnecessary and problematic (because how do you save the resultant int mapping, when the loading code expects that it's reading strings?).  This makes it so that `TokenIndexers` can bypass the vocabulary if what they're given is already an `int`.

I also fixed a small bug in the `SquadReader`, where we were removing newlines.  This was necessary for our old data code where we output one instance per line, but is no longer necessary.

### CODE REVIEW

File: allennlp/data/token_indexers/single_id_token_indexer.py [hunk 1]
```diff
@@ -29,14 +29,19 @@ def __init__(self, namespace: str = 'tokens', lowercase_tokens: bool = False) ->
 
     @overrides
     def count_vocab_items(self, token: str, counter: Dict[str, Dict[str, int]]):
```
Comment (DeNeutoy) [inline allennlp/data/token_indexers/single_id_token_indexer.py]: Can you change the type annotations here or/and add a comment describing in what case the `SingleIdTokenIndexer` receives ints.

File: allennlp/data/token_indexers/token_characters_indexer.py [hunk 2]
```diff
@@ -36,13 +36,18 @@ def __init__(self,
     @overrides
     def count_vocab_items(self, token: str, counter: Dict[str, Dict[str, int]]):
         for character in self.character_tokenizer.tokenize(token)[0]:
-            counter[self.namespace][character] += 1
```
Comment (DeNeutoy) [inline allennlp/data/token_indexers/token_characters_indexer.py]: Same here.

File: allennlp/data/token_indexers/single_id_token_indexer.py [hunk 3]
```diff
@@ -29,14 +29,19 @@ def __init__(self, namespace: str = 'tokens', lowercase_tokens: bool = False) ->
...
     def token_to_indices(self, token: str, vocabulary: Vocabulary) -> int:
-        if self.lowercase_tokens:
-            token = token.lower()
+        if isinstance(token, int):
+            index = token
+        else:
+            if self.lowercase_tokens:
+                token = token.lower()
+            index = vocabulary.get_token_index(token, self.namespace)
         return vocabulary.get_token_index(token, self.namespace)
```
Comment (DeNeutoy) [inline allennlp/data/token_indexers/single_id_token_indexer.py]: I think you need to return `index` here.

Comment (matt-gardner) [reply allennlp/data/token_indexers/single_id_token_indexer.py]: Thanks.

**Pull request merged**